pkg/install/quota.go

```diff
@@ -0,0 +1,83 @@
+package install
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/compute"
+)
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+//checkQuota checks usage quotas vs. resources required by cluster before cluster creation
+func (i *Installer) checkQuota(ctx context.Context) error {
+	//prepare usage client
+	spp := &i.doc.OpenShiftCluster.Properties.ServicePrincipalProfile
+	r, err := azure.ParseResourceID(i.doc.OpenShiftCluster.ID)
+	if err != nil {
+		return err
+	}
+	conf := auth.NewClientCredentialsConfig(spp.ClientID, string(spp.ClientSecret), spp.TenantID)
+	authorizer, err := conf.Authorizer()
+	if err != nil {
+		return err
+	}
+	usageClient := compute.NewUsageClient(r.SubscriptionID, authorizer)
+	return i.calculateAndCompareQuota(ctx, usageClient)
+}
+
+func (i *Installer) calculateAndCompareQuota(ctx context.Context, uc compute.UsageClient) error {
+	//master nodes are already included in the declaration below
+	requiredResources := map[string]int{
+		"cores":              24,
+		"virtualMachines":    3,
+		"standardDSv3Family": 3,
+		"PremiumDiskCount":   3,
+	}
+	if api.VMSizeStandardD8sV3 != i.doc.OpenShiftCluster.Properties.MasterProfile.VMSize {
+		//if master node type changed error out
+		return fmt.Errorf("Unexpected Master node VMSize %s, expected %s", i.doc.OpenShiftCluster.Properties.MasterProfile.VMSize, api.VMSizeStandardD8sV3)
+	}
+	//worker node resource calculation
+	for _, w := range i.doc.OpenShiftCluster.Properties.WorkerProfiles {
+		i.log.Infof("WorkerProfile VMSize %s\n", w.VMSize)
+		requiredResources["virtualMachines"] = requiredResources["virtualMachines"] + w.Count
+		requiredResources["PremiumDiskCount"] = requiredResources["PremiumDiskCount"] + w.Count
+		switch w.VMSize {
+		case api.VMSizeStandardD2sV3:
+			requiredResources["standardDSv3Family"] = requiredResources["standardDSv3Family"] + w.Count
+			requiredResources["cores"] = requiredResources["cores"] + (w.Count * 2)
```

Comment:
I'm not 100% sure that hard coding number of cores here is ok. There is an API that returns SKUs for a lot of things including VMs. You can get info on the number of cores for each VM SKU **per location**. Dispite the fact that numbers semeed to be consistent across locations when I was looking into it last year, I think we should be careful here.

Things to ask ourselves (and probably MSFT): Is there a reason why this data is dynamically served via API and devided by location? Is it becasue amount of resources can change over the time? Is it possible that one size can have different amout of cores/resources in different locations? Or is it possible that some regions do not have some sizes?

Reply:
To summarise: using SKU API will add to the amount of code. If we can avoid using it - great. But relying on data from that API will give us most accurate results.

Reply:
Maybe const map with these details so we can codify them in single location/

Reply:
I asked @Makdaam to start with a hard-coded map, we can extend later if needs be.  But yes, please pull this logic out into a free-standing function.